src/Workspaces/CSharp/Portable/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -605,9 +605,129 @@ private IEnumerable<ITypeSymbol> InferTypeInAttributeArgument(int index, IEnumer
 
             private IEnumerable<ITypeSymbol> InferTypeInArgument(int index, IEnumerable<IMethodSymbol> methods, ArgumentSyntax argumentOpt)
             {
+                if (argumentOpt != null)
+                {
+                    var invocation = argumentOpt?.Parent?.Parent as InvocationExpressionSyntax;
+                    if (invocation != null)
+                    {
+                        // We're trying to figure out the signature of a method we're an argument to. 
+                        // That method may be generic, and we might end up using one of its generic
+                        // type parameters in the type we infer.  First, let's see if we can instantiate
+                        // the methods so that the type can be inferred better.
+                        var invocationTypes = this.InferTypes(invocation).WhereNotNull().ToList();
+                        var instantiatedMethods = methods.Select(m => Instantiate(m, invocationTypes)).ToList();
+
+                        // Now that we've instantiated the methods, filter down to the ones that 
+                        // will actually return a viable type given where this invocation expression
+                        // is.
+                        var filteredMethods = instantiatedMethods.Where(m =>
+                            invocationTypes.Any(t => Compilation.ClassifyConversion(m.ReturnType, t).IsImplicit)).ToList();
+
+                        // If we filtered down to nothing, then just fall back to the instantiated list.
+                        // this is a best effort after all.
+                        methods = filteredMethods.Any() ? filteredMethods : instantiatedMethods;
+                    }
+                }
+
                 return InferTypeInArgument(index, methods.Select(m => m.Parameters), argumentOpt);
             }
 
+            private IMethodSymbol Instantiate(IMethodSymbol method, IList<ITypeSymbol> invocationTypes)
+            {
+                // No need to instantiate if this isn't a generic method.
+                if (method.TypeArguments.Length == 0)
+                {
+                    return method;
+                }
+
+                // Can't infer the type parameters if this method doesn't have a return type.
```

Comment:
As we discussed offline, we infer the type parameters from other parameters elsewhere in the type inferrer, so this comment is true within the context of what we're specifically trying to do here, which is flow from the return type to the parameters. Can you add that clarification to this comment, or more preferably to the comments in InferTypeInArgument above?